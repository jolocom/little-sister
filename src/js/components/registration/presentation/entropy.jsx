import React from 'react'
import Radium from 'radium'
import MaskedImage from './masked-image'
import {RaisedButton} from 'material-ui'

import {Container, Footer} from '../../structure'

import {theme} from 'styles'

const IMAGE_DATA_URL = 'img/entropy.jpg'

const STYLES = {
  container: {
    backgroundImage: 'url(img/img_seedcreation.svg)',
    backgroundPosition: 'center',
    backgroundSize: 'cover',
    position: 'relative',
    padding: 0,
    '@media screen and (min-width: 768px)': {
      padding: '0'
    }
  },
  header: {
    position: 'absolute',
    top: '60px',
    width: '100%',
    maxWidth: '320px',
    padding: '0 16px',
    pointerEvents: 'none',
    boxSizing: 'border-box'
  },
  title: {
    color: '#fff',
    fontSize: theme.textStyles.headline.fontSize,
    fontWeight: theme.textStyles.headline.fontWeight,
    margin: 0
  },
  subtitle: {
    color: '#fff',
    fontSize: theme.textStyles.screenHeader.fontSize,
    fontWeight: theme.textStyles.headline.fontWeight,
    margin: 0
  },
  img: {
    alignItems: 'center',
    width: '100%',
    height: '100%'
  },
  footer: {
    position: 'absolute',
    bottom: 0,
    width: 'auto',
    padding: '32px',
    boxSizing: 'border-box'
  }
}

const Entropy = (props) => {
  let header

  if (props.imageUncoveredPaths.length === 0) {
    header = (
      <header style={STYLES.header}>
        <h1 style={STYLES.title}>
          Hi {props.user}, for... <br /><br />
          <span style={STYLES.subtitle}>
            ...more security we need some random data.<br /><br />
          Please put your finger anywhere on the screen and draw on it randomly.
          </span>
        </h1>
      </header>
    )
  }

  return (
    <Container
      style={STYLES.container}
    >

      {header}

      <MaskedImage
        image={IMAGE_DATA_URL}
        style={STYLES.img}
        maskColor={theme.palette.textColor}
        uncoveredPaths={props.imageUncoveredPaths}
        uncovering={props.imageUncovering}
        onPointUncovered={props.onImagePointUncoverd}
        onUncoveringChange={props.onImageUncoveringChange}
      />

      <Footer style={STYLES.footer}>
        <RaisedButton
          label="NEXT STEP"
          disabled={!props.valid}
          secondary
          onClick={props.onSubmit} />
      </Footer>
    </Container>
  )
}

Entropy.propTypes = {
  imageUncovering: React.PropTypes.bool.isRequired,
  imageUncoveredPaths: React.PropTypes.any,
  user: React.PropTypes.string,
  onImagePointUncoverd: React.PropTypes.func.isRequired,
  onImageUncoveringChange: React.PropTypes.func.isRequired,
  onSubmit: React.PropTypes.func.isRequired,
  valid: React.PropTypes.bool
}

export default Radium(Entropy)
