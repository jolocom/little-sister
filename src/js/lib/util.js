import url from 'url'

// Misc utility functions
let Util = {
  stringLessThan(s1, s2){
    if(s1 < s2) return true
    return false
  },

  stringMin(s1, s2){
    if(Util.stringLessThan(s1, s2)) return s1
    return s2
  },

  stringMax(s1, s2){
    if(Util.stringLessThan(s1, s2)) return s2
    return s1
  },

  distance(x1, y1, x2, y2){
    return Math.sqrt(Math.pow((x1 - x2), 2) +
                     Math.pow((y1 - y2), 2))
  },

  // for short randomStrings
  randomString(length) {
    // Hope this is trully random, will need to do some extra research.
    return Date.now().toString(36).substring(2, length + 3)
  },

  webidRoot(webid) {
    return webid.match(/^(.*)\/profile\/card#me$/)[1]
  },

  urlWithoutHash(target) {
    let obj = url.parse(target)
    obj.hash = null
    return url.format(obj)
  },

  linkToState(target, property) {
    return (e) => {
      target.setState({
        [property]: e.target.value
      })
    }
  }
}

export default Util
