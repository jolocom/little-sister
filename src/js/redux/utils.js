import _ from 'lodash'
import { bindActionCreators } from 'redux'
import { connect as reduxConnect } from 'react-redux'

/**
 * Abstraction on top of connect() from react-redux.
 *
 * In order to reduce boilerplate connecting components to Redux
 * and to prevent tightly coupling them to Redux itself, this decorator
 * is a drop-in replacement that allows you to specify the data and actions
 * you want to access declaritively.
 *
 * Suppose you have a Redux module named widget, exposing its data on
 * the lop-level global state, and with an action named foo, you could use
 * this like:
 *
 * connect({props: ['widget'], actions: ['widget:foo']})(
 *  props => <div onClick={props.foo}>
 *   {props.foo.data}
 *  </div>
 * )
 */
export function connect(params, wantedActions = []) {
  let wantedProps = params.props || params
  if (!wantedProps.map) {
    wantedProps = []
  }
  wantedActions = params.actions || wantedActions

  return reduxConnect(
    (state, props) => {
      if (typeof wantedProps !== 'function') {
        return _.fromPairs(wantedProps.map(prop => {
          const pair = [prop.split('.').slice(-1), state.getIn(prop)]
          if (pair[1].toJS) {
            pair[1] = pair[1].toJS()
          }
        }))
      } else {
        return wantedProps(state, props)
      }
    },
    (dispatch, props) => {
      if (typeof wantedActions !== 'function') {
        return bindActionCreators(_.fromPairs(wantedActions.map(id => {
          const [moduleName, actionName] = id.split(':')
          const module = require('redux/modules/' + moduleName)
          return [actionName, [actionName, module[actionName]]]
        })), dispatch)
      } else {
        return wantedActions(dispatch, props)
      }
    }
  )
}
