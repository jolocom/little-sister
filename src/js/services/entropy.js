import sjcl from 'sjcl'

export default class EntropyService {
  constructor() {
    this.generator = new sjcl.prng(5) // eslint-disable-line new-cap
    // this.generator.startCollectors()
  }

  addFromDelta(d) {
    this.generator.addEntropy(d, 1, 'mouse')
  }

  getProgress() {
    return this.generator.getProgress()
  }

  isReady() {
    return this.generator.isReady()
  }

  getRandomNumbers(count) {
    return this.generator.randomWords(count).map(s => parseInt(s))
  }

  getRandomString(wordCount) {
    return this.generator.randomWords(wordCount).join('')
  }
}
