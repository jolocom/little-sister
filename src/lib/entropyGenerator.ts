const sjcl = require('sjcl')

export interface EntropyGeneratorInterface {
  addFromDelta: (d: number) => void;
  getProgress: () => number;
  generateRandomString: (wordCount: number) => string;
}

export class EntropyGenerator implements EntropyGenerator {
  private generator = new sjcl.prng(10)

  addFromDelta(d: number) : void {
    this.generator.addEntropy(d, 1, 'user')
  }

  getProgress() : number {
    return this.generator.getProgress()
  }

  generateRandomString(wordCount: number) : string {
    // returns an array of length wordCount filled with random 4 byte words.
    const intArray = new Int32Array(this.generator.randomWords(wordCount))
    const buf = Buffer.from(intArray.buffer)
    return buf.toString('hex')
  }
}
