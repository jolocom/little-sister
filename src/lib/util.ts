import { claimsMetadata, JolocomLib } from 'jolocom-lib'
import {
  uiCategoryByCredentialType,
  Categories,
  uiCredentialTypeByType,
} from './categories'
import { BaseMetadata } from 'cred-types-jolocom-core'
import { BackendMiddleware } from 'src/backendMiddleware'

export const getClaimMetadataByCredentialType = (
  type: string,
): BaseMetadata => {
  const uiType = Object.keys(uiCredentialTypeByType).find(
    item => uiCredentialTypeByType[item] === type,
  )

  const relevantType = Object.keys(claimsMetadata).find(
    key => claimsMetadata[key].type[1] === uiType,
  )

  if (!relevantType) {
    throw new Error("Unknown credential type, can't find metadata")
  }

  return claimsMetadata[relevantType]
}

export const getUiCredentialTypeByType = (type: string[]): string =>
  uiCredentialTypeByType[type[1]] || prepareLabel(type[1])

export const getCredentialUiCategory = (type: string): string => {
  const uiCategories = Object.keys(uiCategoryByCredentialType)

  const category = uiCategories.find(uiCategory => {
    const categoryDefinition = uiCategoryByCredentialType[uiCategory]
    const credentialFitsDefinition = categoryDefinition.some(
      entry => entry === type,
    )
    return credentialFitsDefinition
  })

  return category || Categories.Other
}

export const areCredTypesEqual = (first: string[], second: string[]): boolean =>
  first.every((el, index) => el === second[index])

export const prepareLabel = (label: string): string => {
  const words = label.split(/(?=[A-Z0-9])/)
  return words.length > 1 ? words.map(capitalize).join(' ') : capitalize(label)
}

export const capitalize = (word: string): string =>
  `${word[0].toUpperCase()}${word.slice(1)}`

export const compareDates = (date1: Date, date2: Date): number =>
  Math.floor(
    (Date.UTC(date2.getFullYear(), date2.getMonth(), date2.getDate()) -
      Date.UTC(date1.getFullYear(), date1.getMonth(), date1.getDate())) /
      (1000 * 60 * 60 * 24),
  )

export const instantiateIdentityWallet = async (
  backendMiddleware: BackendMiddleware,
) => {
  const { keyChainLib, storageLib, encryptionLib } = backendMiddleware

  const password = await keyChainLib.getPassword()
  const decryptedSeed = encryptionLib.decryptWithPass({
    cipher: await storageLib.get.encryptedSeed(),
    pass: password,
  })

  // TODO: rework the seed param on lib, currently cleartext seed is being passed around. Bad.
  const userVault = new JolocomLib.KeyProvider(
    Buffer.from(decryptedSeed, 'hex'),
    password,
  )
  return await backendMiddleware.setIdentityWallet(userVault, password)
}
