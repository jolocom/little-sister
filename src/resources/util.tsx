import React from 'react'
import { CredentialTypes } from 'src/lib/categories'
import { EmailIcon, NameIcon, PhoneIcon, AccessibilityIcon } from '.'

export const getCredentialIconByType = (type: string) => {
  const typeToIconMap = {
    [CredentialTypes.Name]: <NameIcon />,
    [CredentialTypes.Email]: <EmailIcon />,
    [CredentialTypes.MobilePhone]: <PhoneIcon />,
  }

  return typeToIconMap[type] || <AccessibilityIcon />
}
