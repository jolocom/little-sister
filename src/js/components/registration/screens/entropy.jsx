import React from 'react'
import { connect } from 'redux/utils'
import ImageMaskBuilder from 'lib/image-mask-builder'
import Presentation from '../presentation/entropy'

@connect({
  props: ['registration'],
  actions: [
    'registration:goForward',
    'registration:setMaskedImageUncovering',
    'registration:addEntropyFromDeltas'
  ],
  pure: false
})
export default class RegistrationEntropyScreen extends React.Component {
  static propTypes = {
    registration: React.PropTypes.object.isRequired,

    goForward: React.PropTypes.func.isRequired,
    addEntropyFromDeltas: React.PropTypes.func.isRequired,
    setMaskedImageUncovering: React.PropTypes.func.isRequired
  }

  constructor() {
    super()

    this.state = {
      count: 0,
      imageMask: new ImageMaskBuilder()
    }
  }

  handleUncoveringChange(uncovering) {
    this.props.setMaskedImageUncovering(uncovering)

    if (uncovering) {
      this.state.imageMask.startNewPath()
    } else {
      this.state.imageMask.endPath()
    }
  }

  handleUncoveredPoint(x, y) {
    this.state.imageMask.addPoint(x, y)
    this.setState({count: this.state.count + 1})
  }

  render() {
    return <Presentation
      imageUncoveredPaths={this.state.imageMask.paths}
      imageUncovering={this.props.registration.maskedImage.uncovering}
      valid={this.props.registration.passphrase.sufficientEntropy}
      user={this.props.registration.humanName.value}
      onImagePointUncoverd={(...args) => this.handleUncoveredPoint(...args)}
      onImageUncoveringChange={(val) => this.handleUncoveringChange(val)}

      onSubmit={this.props.goForward}
    />
  }
}
