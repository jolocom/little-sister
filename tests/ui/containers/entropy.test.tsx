import React from 'react'
import { shallow } from 'enzyme'
import { EntropyContainer } from 'src/ui/registration/containers/entropy'
import { EntropyGenerator } from 'src/lib/entropyGenerator'
import * as util from '@jolocom/sdk/js/util'
import { stub, reveal } from 'tests/utils'

describe('Entropy container', () => {
  const mockInstanceGenerator = (
    instance: EntropyContainer,
    progressValues = [0],
  ) => {
    const getProgress = jest.fn().mockReturnValue(progressValues[0])
    // @ts-ignore private
    const generatorMock = (instance.entropyGenerator = stub<EntropyGenerator>({
      getProgress,
      addFromDelta: jest.fn(),
      generateRandomString: jest.fn().mockReturnValue('randomString'),
    }))

    progressValues.forEach(v => getProgress.mockReturnValueOnce(v))

    return generatorMock
  }

  const props = {
    submit: jest.fn(),
  }

  it('mounts correctly and matches the snapshot', () => {
    const rendered = shallow(<EntropyContainer {...props} />)
    expect(rendered).toMatchSnapshot()
  })

  it('correctly creates an entropy generator upon component mount', () => {
    const entropySpy = jest.spyOn(
      EntropyContainer.prototype,
      // @ts-ignore private
      'setUpEntropyGenerator',
    )
    shallow(<EntropyContainer {...props} />)
    expect(entropySpy).toHaveBeenCalledTimes(1)
  })

  it('correctly handles added points with the entropy generator', () => {
    const rendered = shallow(<EntropyContainer {...props} />)
    const instance = rendered.instance() as EntropyContainer
    const generatorMock = mockInstanceGenerator(instance)

    expect(rendered.state()).toMatchSnapshot()

    // @ts-ignore private
    instance.updateEntropyProgress = jest.fn()
    // @ts-ignore private
    instance.addPoint(150, 220)
    // @ts-ignore private
    expect(instance.updateEntropyProgress).toHaveBeenCalledTimes(1)

    expect(generatorMock.getProgress).toHaveBeenCalledTimes(1)
    expect(generatorMock.addFromDelta).toHaveBeenCalledTimes(2)
    expect(reveal(generatorMock).addFromDelta.mock.calls).toEqual([
      [150],
      [220],
    ])

    expect(rendered.state()).toMatchSnapshot()
  })

  it('correctly triggers a random string generation when there is sufficient entropy', async () => {
    const generateSecureRandomBytesMock = jest
      .spyOn(util, 'generateSecureRandomBytes')
      .mockResolvedValue(Promise.resolve(Buffer.from('moreEntropy')))

    const rendered = shallow(<EntropyContainer {...props} />)
    const instance = rendered.instance() as EntropyContainer
    mockInstanceGenerator(instance, [0.9, 1])

    expect(rendered.state()).toMatchSnapshot()

    rendered.setState({ entropyProgress: 1 })
    // @ts-ignore private
    await instance.updateEntropyProgress()

    expect(generateSecureRandomBytesMock).toHaveBeenCalled()
    expect(rendered.state()).toMatchSnapshot()
  })

  it('does not trigger a random string generation when there is not sufficient entropy', async () => {
    const rendered = shallow(<EntropyContainer {...props} />)
    const instance = rendered.instance() as EntropyContainer
    mockInstanceGenerator(instance)

    expect(rendered.state()).toMatchSnapshot()

    rendered.setState({ entropyProgress: 0.99 })
    // @ts-ignore private
    await instance.updateEntropyProgress()

    expect(rendered.state()).toMatchSnapshot()
  })
})
