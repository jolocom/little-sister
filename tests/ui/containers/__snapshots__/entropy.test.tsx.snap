// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entropy container correctly handles added points with the entropy generator 1`] = `
Object {
  "encodedEntropy": "",
  "entropyProgress": 0,
  "isDrawn": false,
  "sufficientEntropy": false,
}
`;

exports[`Entropy container correctly handles added points with the entropy generator 2`] = `
Object {
  "encodedEntropy": "",
  "entropyProgress": 1,
  "isDrawn": false,
  "sufficientEntropy": false,
}
`;

exports[`Entropy container correctly triggers a random string generation when there is sufficient entropy 1`] = `
Object {
  "encodedEntropy": "",
  "entropyProgress": 0,
  "isDrawn": false,
  "sufficientEntropy": false,
}
`;

exports[`Entropy container correctly triggers a random string generation when there is sufficient entropy 2`] = `
Object {
  "encodedEntropy": "randomString",
  "entropyProgress": 1,
  "isDrawn": false,
  "sufficientEntropy": true,
}
`;

exports[`Entropy container does not trigger a random string generation when there is not sufficient entropy 1`] = `
Object {
  "encodedEntropy": "",
  "entropyProgress": 0,
  "isDrawn": false,
  "sufficientEntropy": false,
}
`;

exports[`Entropy container does not trigger a random string generation when there is not sufficient entropy 2`] = `
Object {
  "encodedEntropy": "",
  "entropyProgress": 0.99,
  "isDrawn": false,
  "sufficientEntropy": false,
}
`;

exports[`Entropy container mounts correctly and matches the snapshot 1`] = `
<EntropyComponent
  addPoint={[Function]}
  progress={0}
  submitEntropy={[Function]}
/>
`;
